libtock/util/streaming_process_slice.h

```diff
@@ -0,0 +1,85 @@
+#include "../tock.h"
+
+typedef struct {
+  uint8_t* app_buffer_ptr;
+  size_t app_buffer_size;
+  uint32_t driver;
+  uint32_t allow;
+} streaming_process_slice_state_t;
+
+// Initialize a "streaming process slice" read-write allow slot
+//
+// This method allows a userspace buffer into a "streaming process
+// slice" allow slot, implementing its atomic-swap semantics and
+// header layout. The streaming process slice abstraction allows a
+// userspace process to lossessly receive data from a kernel
+// capsule. This is done by maintaining two buffers, where at any time
+// one of which is owned by the kernel (for writing new, incoming data
+// into) and one by the application, to process received data. These
+// buffers are atomically swapped by the application, upon receipt of
+// a signal that some data has been inserted into the kernel-owned
+// buffer (such as an upcall).
+//
+// This method abstracts this interface by consuming one buffer and
+// splitting it into two halves, owned by the application and kernel
+// respectively. It tracks all necessary state in the
+// `streaming_process_slice_state_t` object. For this to work, the
+// passed buffer must be able to hold at least two streaming process
+// slice headers (8 byte each), i.e., it must be at least 16 bytes
+// long.
```

Comment:
Why not have the user provide two separate buffer pointers? To me, it would make sense for an application to implement two buffers of the same size (Header + Data) and pass both into this, rather than one buffer of double size (2x Header + Data).

Reply:
Yeah, that's also an option. I pondered over this for a while, but didn't manage to come up with any good arguments for this arguably more complex interface. There's not really a good reason for why an app would want to allow two differently sized buffers (as they're both used for the same purpose), and I don't know whether there are any significant benefits to being able to allocate the two halves separately.

I'm open to changing this, but wouldn't want to do so when there's not a good reason. I quite liked the elegance of this behaving quite similar to a regular allow, apart from the need to carry the state struct in user code.

Reply:
My argument would be that it's currently hidden that two buffers are happening, so it's easy to incorrectly calculate how much space the buffer should have to support that. For example, needing to reserve 2*header space.

In reality, there should be library interfaces that use this and provide something better anyways. This is sort of an internal, developer interface. So it's probably not a big deal to trust them to read and understand documentation.

Reply:
Hm, to me there's not much of a difference to telling users of this interface that they need to reserve `2*HEADER_LEN` vs. `HEADER_LEN` in the buffer, or passing two 8-byte vs. one 16-byte buffer.

In line with your argument, if we expect a single buffer, than any higher level abstraction must also take a single buffer. If we were to take two, then the higher-level layer could still do the splitting internally...

I'm still torn on this. I think the current interface is neat, and I'm yet to be convinced that there will ever be a time where we'd want to pass two non-contiguous buffers.

Reply:
Here's a confusion: let's say you set `#define HEADER_LEN (16)`. Somebody could reserve 2*HEADER_LEN still. Similarly if you set `#define HEADER_LEN (8)` somebody could only reserve 1*HEADER_LEN. But again, these are really end-user arguments, and internal developers can be expected to try harder. Our documentation can be clear on which is correct, and most users will probably be a member of the core team anyways. Maybe in the long term there will be others, but they can copy from our examples.

I do NOT have a good example of why the buffers would need to be non-contiguous or different in size.

So, I'd say leave this stream open and see if anyone else agrees with one side or the other. If there aren't a lot of people on the "two buffers" side, I'd say leave it as-is.

Reply:
I switched it to using two buffers.

Reply:
I split this into two commits, so we can refer back to the previous version for how I implemented the splitting (it's pretty trivial, but still).